### ISSUE TITLE
HierarchicalMatrices should use BlockArrays

### ISSUE THREAD

**Issue by dlfivefifty** [MikaelSlevinsky/HierarchicalMatrices.jl#4]
BlockArrays has a well-developed block matrix structure, so this would avoid needing your own `Block`.

**Comment by MikaelSlevinsky** [comment]
You're talking mainly about the abstract block array interface, and not the concrete block array type, right? AFAICT, hierarchical matrices are block arrays where the blocks are different types, but this is staged for performance. On the other hand, the concrete block array type has templated the block type. I don't see how a `getindex` on the blocks would ever be performant for hierarchical matrices due to the type instability. `setindex!`, though is type stable.

**Comment by dlfivefifty** [comment]
That's not correct: `AbstractBlockMatrix{T}` only incorporates the type of the entries `T`, not of the blocks.
